api/src/controllers/translations.ts

```diff
@@ -0,0 +1,95 @@
+import express from 'express';
+import { ForbiddenException } from '../exceptions/index.js';
+import { respond } from '../middleware/respond.js';
+import { validateBatch } from '../middleware/validate-batch.js';
+import { TranslationsService } from '../services/index.js';
+import asyncHandler from '../utils/async-handler.js';
+import type { PrimaryKey } from '@directus/types';
+
+const router = express.Router();
+
+const readHandler = asyncHandler(async (req, res, next) => {
+	const service = new TranslationsService({
+		accountability: req.accountability,
+		schema: req.schema,
+	});
+
+	const result = await service.readByQuery(req.sanitizedQuery);
+
+	res.locals['payload'] = {
+		data: result,
+	};
+
+	return next();
+});
+
+router.search('/', validateBatch('read'), readHandler, respond);
+router.get('/', readHandler, respond);
+
+const upsertHandler = asyncHandler(async (req, res, next) => {
+	const service = new TranslationsService({
+		accountability: req.accountability,
+		schema: req.schema,
+	});
+
+	try {
+		let savedKeys: PrimaryKey[] = [];
+
+		const existingKeys: PrimaryKey[] = (
+			await service.readByQuery({
+				fields: ['id'],
+				filter: {
+					key: { _eq: req.params['key']! },
+				},
+			})
+		).map(({ id }) => id);
+
+		if (Array.isArray(req.body)) {
+			savedKeys = await service.upsertMany(req.body);
+			const result = await service.readMany(savedKeys, req.sanitizedQuery);
+			res.locals['payload'] = { data: result || null };
+		} else {
+			const savedKey = await service.upsertOne(req.body);
+			savedKeys.push(savedKey);
+			const result = await service.readOne(savedKey, req.sanitizedQuery);
+			res.locals['payload'] = { data: result || null };
+		}
+
+		const deleteKeys = existingKeys.filter((id) => !savedKeys.includes(id));
+		if (deleteKeys.length > 0) await service.deleteMany(deleteKeys);
+	} catch (error: any) {
+		if (error instanceof ForbiddenException) {
+			return next();
+		}
+
+		throw error;
+	}
+
+	return next();
+});
+
+router.post('/:key', upsertHandler, respond);
+router.patch('/:key', upsertHandler, respond);
```

Comment:
I like that idea 🤔 However i dont think we are using PUT yet right? perhaps we should introduce that with a broader upsert for items collections feature like https://github.com/directus/directus/discussions/5706 🤔

Reply:
Not as far as I know, no! Adding upsertion to the API in general is definitely a bigger task that needs a lot more planning (as @rijkvanzanten mentioned in the discussion you linked, it is not as easy as always using `PUT` for upserts because it indicates _replace_ instead of _update_). I'm just a little worried that the translations controller will be the only outlier that works differently once we add upsertion to all the other controllers.

Reply:
translations is already an outlier as the `key` here isn't the primary key. 👀

Reply:
`/translations` should be the same as any other endpoint 👍🏻

Reply:
> translations is already an outlier as the `key` here isn't the primary key. 👀

You're not wrong! We do have controllers that use different routes than the items controller , though, but they at least all flow the same rules for `POST`/`PATCH`/etc.

Making `/translations` act the same as `/items` would indeed be best 👍